### DISCUSSION TITLE
Allow magic link login even if the email is enrolled with an SSO provider

### DISCUSSION THREAD

**Discussion by vpukhanov** [outline/outline#3660] [Ideas]
Hey folks.

At our company we are self-hosting Outline with Slack and Magic Link authentication providers enabled. We're currently migrating off of Slack, but would like for both auth methods to be available in the interim.

Our problem was that if the email was enrolled in Outline via Slack provider, even if the user selects the email authentication, they get redirected to Slack to authenticate there. Here's the code that enforces this behavior: https://github.com/outline/outline/blob/9faa5dd01172d501e0467372fefdabe6b3186a0d/server/routes/auth/providers/email.ts#L77-L87

We have changed this in our fork of Outline, and we would be interested in bringing this change to upstream. I propose introducing a new environment variable `PREFER_SSO_OVER_EMAIL_AUTH` with the default value of `true` to keep the current behavior intact. If this is set to false, Outline would allow signing in via email even if it is enrolled with an SSO provider.

Would you be interested in accepting a pull request with this change?

**Comment by tommoor** [comment]
Hi @vpukhanov – this logic is very purposeful as allowing magic link auth for an SSO account could be used as a security loophole to get around 2FA requirements that may be in place.

Given that this would introduce rarely-used additional complexity to the authentication flow we're not interested in a patch at this time, but I'm glad you were able to make this work for your team.

**Reply by vpukhanov** [reply]
Got you, and thanks for explaining the reasoning behind it

**Reply by gouku** [reply]
@tommoor We have the same issue here. Our Outline is deployed behind the firewall with whitelist internet access so the security should not be a problem. But our OAuth does not always work as expected so we'd like to try the magic link method. In this situation, we got a login loophole that no one can login until we fix the OAuth provider.

An option like `PREFER_SSO_OVER_EMAIL_AUTH` could really helpful in these cases.

**Reply by tommoor** [reply]
We're spending a lot of time in this area at the moment internally, rejigging how authentication is handled and matched so will definitely bear this in mind 👍 cc @thenanyu
